src/core/net/ip6.cpp

```diff
@@ -151,6 +151,58 @@ ThreadError Ip6::AddMplOption(Message &message, Header &header, IpProto nextHead
     return error;
 }
 
+ThreadError Ip6::InsertMplOption(Message &aMessage)
+{
+    ThreadError error = kThreadError_None;
+    Header ip6header;
+
+    aMessage.Read(0, sizeof(ip6header), &ip6header);
+    VerifyOrExit(ip6header.GetDestination().IsRealmLocalMulticast(),);
+
+    aMessage.RemoveHeader(sizeof(ip6header));
+
+    if (ip6header.GetNextHeader() == kProtoHopOpts)
+    {
+        HopByHopHeader hbh;
+        uint8_t hbhLength = 0;
+        OptionMpl mplOption;
+
+        // read existing hop-by-hop option header
+        aMessage.Read(0, sizeof(hbh), &hbh);
+        hbhLength = (hbh.GetLength() + 1) * 8;
+
+        // increase existing hop-by-hop option header length by 8 bytes
+        hbh.SetLength(hbh.GetLength() + 1);
+        aMessage.Write(0, sizeof(hbh), &hbh);
+
+        // make space for MPL Option + padding by shifting hop-by-hop option header
+        aMessage.Prepend(NULL, 8);
+        aMessage.CopyTo(0, 8, 8, aMessage);
+
+        // insert MPL Option
+        mMpl.InitOption(mplOption, ip6header.GetSource());
+        aMessage.Write(hbhLength, mplOption.GetTotalLength(), &mplOption);
+
+        // insert Pad Option (if needed)
+        if (mplOption.GetTotalLength() % 8)
```

Comment:
In this case, the IPv6 Hop-by-Hop Option header is already included when the IPv6 datagram was submitted to us.  As a result, we do not include that in the padding calculation.